### ISSUE TITLE
TransformerEncoder/TransformerDecoder has same initial parameters for all layers

### ISSUE THREAD

**Issue by albertz** [pytorch/pytorch#86274]
### 🐛 Describe the bug

Example code (based on [documentation](https://pytorch.org/docs/stable/generated/torch.nn.TransformerEncoder.html#torch.nn.TransformerEncoder)):
```python
encoder_layer = nn.TransformerEncoderLayer(d_model=512, nhead=8)
transformer_encoder = nn.TransformerEncoder(encoder_layer, num_layers=6)
```
Then you have:
```python
(transformer_encoder.layers[0].linear1.weight == transformer_encoder.layers[1].linear1.weight).all() == True
```
Etc.

This is because:
```python
self.layers = _get_clones(encoder_layer, num_layers)
```
Where `_get_clones` uses `deepcopy`.

The same problem is for `TransformerDecoder`.

This is unexpected behavior to me but maybe it is expected and just not well documented?

Note that `Transformer` itself does not have this problem because it calls `_reset_parameters` after all the layers are assigned.

### Versions

(I have an outdated version but I checked the source code of the current version.)

**Comment by erichan1** [comment]
I think this is expected and not well documented. But it could definitely be changed. Is the desire to make sure the weights are initialized differently between layers?

**Comment by albertz** [comment]
I assume that users usually do not want this. Users of `nn.Transformer` anyway would not have the problem though, and other users might already have overwritten the param init *if they know about this*.

**Comment by erichan1** [comment]
I see. Would you be open to making a PR fixing this? We have a couple of options
1. Add a reset_parameters() function to both TransformerEncoder and TransformerDecoder but do not automatically call it. Let people call it themselves. 
2. Do option 1 but also add a call to reset_parameters() in the init function.
3. No code changes and add documentation that all layers will be initialized the same. 

I was thinking to just do option 1. As a user, what do you think?

**Comment by albertz** [comment]
> Would you be open to making a PR fixing this?

Yes, if I find the time. Not sure exactly when.

> As a user, what do you think?

My main issue here is about unexpected behavior - or rather, it would be nice if a normal use of those modules would yield expected behavior.

The problem is introduced due to the use of `deepcopy`. So maybe it makes sense to fix the param init right after the `deepcopy`, i.e. in the `TransformerEncoder` and `TransformerDecoder`, so basically your option 2.

Then the `_reset_parameters` in the `Transformer` itself might not be needed anymore? I'm not sure if people might pass `custom_encoder` or `custom_decoder` thought which depend on the `Transformer._reset_parameters` though.

**Comment by adosar** [comment]
> I assume that users usually do not want this. Users of `nn.Transformer` anyway would not have the problem though, and other users might already have overwritten the param init _if they know about this_.

I agree, this is in accordance with the user expecting a random initialization of model's parameters.
